dali/operators/decoder/video/video_decoder_base.h

```diff
@@ -18,118 +18,225 @@
+  VideoDecoderBase(const OpSpec &spec) : Operator<Backend>(spec) {
+    auto pad_mode_str = spec_.template GetArgument<std::string>("pad_mode");
+    if (pad_mode_str == "none") {
+      padding_mode_ = PaddingMode::kNone;
+    } else if (pad_mode_str == "constant") {
+      padding_mode_ = PaddingMode::kConstant;
+    } else if (pad_mode_str == "edge") {
+      padding_mode_ = PaddingMode::kEdge;
+    } else {
+      DALI_FAIL(make_string("Invalid pad_mode: ", pad_mode_str));
+    }
+
+    if (padding_mode_ == PaddingMode::kConstant) {
+      pad_value_ = spec_.template GetArgument<int>("pad_value");
+      DALI_ENFORCE(pad_value_ >= 0 && pad_value_ <= 255,
+                   "pad_value must be in range [0, 255]");
+    }
+  }
 
   DISABLE_COPY_MOVE_ASSIGN(VideoDecoderBase);
 
  protected:
-  void ValidateInput(const Workspace &ws) {
+  std::unique_ptr<FramesDecoderImpl> CreateDecoder(const char* data, size_t size, 
+                                                   bool build_index,
+                                                   std::string_view source_info,
+                                                   cudaStream_t stream = 0) {
+    if constexpr (std::is_same_v<Backend, CPUBackend>) {
+      return std::make_unique<FramesDecoderImpl>(data, size, build_index, true, -1, source_info);
+    } else {
+      return std::make_unique<FramesDecoderImpl>(data, size, stream, build_index, -1, source_info);
+    }
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) override {
     const auto &input = ws.Input<InBackend>(0);
+    
+    // Validate input
     DALI_ENFORCE(input.type() == DALI_UINT8,
                  "Type of the input buffer must be uint8.");
     DALI_ENFORCE(input.sample_dim() == 1,
                  "Input buffer must be 1-dimensional.");
     for (int64_t i = 0; i < input.num_samples(); ++i) {
       DALI_ENFORCE(input[i].shape().num_elements() > 0,
                    make_string("Incorrect sample at position: ", i, ". ",
-                               "Video decoder does not support empty input samples."));
+                             "Video decoder does not support empty input samples."));
     }
-  }
 
-
-  TensorListShape<4> ReadOutputShape() {
-    TensorListShape<4> shape(frames_decoders_.size());
-    for (size_t s = 0; s < frames_decoders_.size(); ++s) {
-      TensorShape<4> sample_shape;
-      sample_shape[0] = frames_decoders_[s]->NumFrames();
+    int batch_size = input.num_samples();
+
+    // Get frame selection parameters
+    if (start_frame_.HasValue())
+      start_frame_.Acquire(spec_, ws, batch_size);
+    if (stride_.HasValue())
+      stride_.Acquire(spec_, ws, batch_size);
+    if (sequence_length_.HasValue())
+      sequence_length_.Acquire(spec_, ws, batch_size);
+
+    frames_decoders_.resize(batch_size);
+
+    bool build_index = start_frame_.HasValue() ||
+                      stride_.HasValue() ||
+                      sequence_length_.HasValue();
```

Comment:
I think we can provide more fain grained control. build_index is true only if start_frame_ is >0, other wise we don't need that. Index is used (now) to seek to the first frames, and then we decode one after another without index (so presence of stride of sequence length doesn't make any difference). Long term we may decide to seek to next frame if it is past another key frame so we don't have to decode that many frames in between, but for now I we probably don't need that.

Reply:
Will do